docs/zkapps/tutorials/09-recursion.mdx

```diff
@@ -0,0 +1,312 @@
+
+To start, let’s set up our ZkProgram:
+
+```typescript
+class RollupState extends Struct({
+  initialRoot: Field,
+  latestRoot: Field,
+}) {
+  static createOneStep(...)
+
+  static createMerged(state1: RollupState, state2: RollupState) {
+	  return new RollupState({
+      initialRoot: state1.initialRoot,
+      latestRoot: state2.latestRoot
+    });
+  }
+
+  static assertEquals(state1: RollupState, state2: RollupState) {
+    state1.initialRoot.assertEquals(state2.initialRoot);
+    state1.latestRoot.assertEquals(state2.latestRoot);
+  }
+}
+
+const Rollup = Experimental.ZkProgram({
+  publicInput: RollupState,
+
+  methods: {
+	oneStep: {
+		...
+	},
+
+  merge: {
+  	privateInputs: [ SelfProof, SelfProof ],
+
+  	method(
+    	newState: RollupState,
+    	rollup1proof: SelfProof<RollupState>,
+    	rollup2proof: SelfProof<RollupState>,
+  	) {
+      rollup1proof.verify(); // A -> B
+      rollup2proof.verify(); // B -> C
+
+      rollup1proof.publicInput.initialRoot.assertEquals(newState.initialRoot);
+
+      rollup1proof.publicInput.latestRoot.assertEquals(rollup2proof.publicInput.initialRoot);
+
+      rollup2proof.publicInput.latestRoot.assertEquals(newState.latestRoot);
+  	}
+}
+```
+
+A proof generated by our `merge` method will indicate there is a valid sequence of transactions (i.e. applications of “oneStep”) that get from an “initialRoot” (the root of a MerkleMap) to a “latestRoot” (root of the Merkle map after transactions are applied).
+
+We will consume these in a SmartContract, which will use the recursive proof to update its internal state:
+
+```typescript
+class RollupContract extends SmartContract {
+  @state(Field) state = State<Field>();
+
+  deploy(args: DeployArgs) {
+    super.deploy(args);
+    this.setPermissions({
+      ...Permissions.default(),
+      editState: Permissions.proofOrSignature(),
+    });
+  }
+
+  @method initStateRoot(stateRoot: Field) {
+    this.state.set(stateRoot);
+  }
+
+  @method update(rollupStateProof: RollupProof) {
+    const currentState = this.state.get();
+    this.state.assertEquals(currentState);
+
+    rollupStateProof.publicInput.initialRoot.assertEquals(currentState);
+
+    rollupStateProof.verify();
+
+    this.state.set(rollupStateProof.publicInput.latestRoot);
+  }
+}
+```
+
+Last, is just to fill in the methods above, particularly, `oneStep()', and `createOneStep()`. This single “step” of the rollup will check the value at an “account” was incremented by a particular amount.
+
+To check that the value at an account was incremented by a particular amount, our code computes an updated state–inside the recursive snark–by calling `RollupState.createOneStep()`. And then, asserting that the new state of the recursive snark, is equivalent to the computed state. Inside `createOneStep()`, a single step of the rollup is created.
```

Comment:
Can you double check that the em dash (—) character is used here. Looks like it might be a hyphen but hard to tell with Github's font